File: packages/raisin-client-js/src/agent-runs.ts
/**
 * Durable agent runs over HTTP.
 *
 * A run is a server-side, crash-safe execution of an agent: a monotonic event
 * log, a lifecycle (`queued → running → waiting/paused → completed/failed/
 * stopped`), and atomic, idempotent controls (stop, pause, resume, steer,
 * approve, answer). Two ways to drive one:
 *
 * - **server-driven** — create it with a `reducer` (any function, in any
 *   supported language); the server's job queue drives it and you watch it
 *   with {@link AgentRunsApi.stream};
 * - **client-driven** — create it without a reducer and drive it yourself with
 *   the lease calls ({@link AgentRunsApi.acquire}, {@link AgentRunsApi.begin},
 *   {@link AgentRunsApi.finish}, ...). Your operations are still durable and
 *   fenced, and anyone authorized can still stop or steer the run.
 */

import {
  checkpointBody,
  spawnBody,
  type ChildAction,
  type CheckpointOptions,
  type MailboxEntry,
  type SpawnChildOptions,
  type SpawnChildResult,
} from './agent-run-children';

/** Where a run lives, what it is about, and what it runs as. */
export interface CreateAgentRunOptions {
  /** What the run is about (one live run per subject). */
  subject: { workspace: string; path: string; node_id?: string | null };
  /** Branch (default "main"). */
  branch?: string;
  /** Opaque agent reference, e.g. an agent node path. */
  agentRef?: string;
  /** Idempotency key for the create (a retried create returns the same run). */
  createKey?: string;
  /** Limits; exceeding one pauses (default) or fails the run. */
  budgets?: Record<string, unknown>;
  /** Input, delivered to the reducer as `run_started.data.input`. */
  input?: unknown;
  /** Server-driven: the reducer function. Omit for a client-driven run. */
  reducer?: { function_path: string; handler?: string };
  /** Opaque executor configuration (e.g. `model_turn_function`). */
  executorConfig?: Record<string, unknown>;
  /** A secret whose holder may control the run. */
  controlCapability?: string;
  /** Run as this agent (`"ws:/path"` or `"/path"`), on your behalf. */
  asAgent?: string;
}

/** The answer to a create. */
export interface CreateAgentRunResult {
  run_id: string;
  /** False when the subject's live run (or the create key's run) came back. */
  created: boolean;
  status: AgentRunStatus;
}

export type AgentRunStatus =
  | 'queued'
  | 'running'
  | 'waiting'
  | 'paused'
  | 'cancelling'
  | 'completed'
  | 'failed'
  | 'stopped';

/** A run as a reader sees it. */
export interface AgentRunView {
  run: Record<string, unknown> & { run_id: string; last_seq: number };
  status: AgentRunStatus;
  /** The plan to display, already overridden by the run status. */
  projection: { items?: Array<{ key: string; title: string; status: string }>; summary?: string } | null;
}

/** One durable event (`seq` is contiguous from 1). */
export interface AgentRunEvent {
  run_id: string;
  seq: number;
  at_ms: number;
  turn?: number | null;
  op_id?: string | null;
  kind: { type: string; [key: string]: unknown };
}

/** The answer to a control. */
export type AgentRunControlAck =
  | { ack: 'applied'; seq: number }
  | { ack: 'duplicate'; original_seq: number }
  | { ack: 'rejected'; reason: string; seq: number };

/** A control command, as the server's contract spells it. */
export type AgentRunCommand =
  | { command: 'stop'; reason?: string }
  | { command: 'pause' }
  | { command: 'resume'; budget_increase?: Record<string, unknown>; accept_reducer_change?: boolean }
  | { command: 'steer'; input: unknown }
  | {
      command: 'approve';
      request_id: string;
      decision: { decision: 'approve' } | { decision: 'reject'; reason?: string };
      subject_digest: string;
    }
  | { command: 'provide_input'; request_id: string; value: unknown };

/** A lease held by a client driver; present it on every driver call. */
export interface AgentRunFence {
  owner: string;
  epoch: number;
}

/** Begin an operation (client-driven runs). */
export interface BeginOperationOptions {
  kind: 'model_turn' | 'tool_call' | 'compaction' | string;
  input?: unknown;
  replay_safe?: boolean;
  non_interruptible?: boolean;
  for_call_id?: string;
  answers?: string[];
}

/** Finish an operation (client-driven runs). */
export interface FinishOperationOptions {
  outcome: 'succeeded' | 'waiting' | 'retryable' | 'blocked' | 'failed' | 'cancelled';
  payload?: unknown;
  tool_calls?: string[];
  usage?: { input_tokens: number; output_tokens: number };
  resume_key?: string;
}

/** How the API reaches the server (supplied by the HTTP client). */
export interface AgentRunsTransport {
  request<T>(method: string, path: string, body?: unknown): Promise<T>;
  /** Absolute URL for `path`. */
  url(path: string): string;
  /** Headers for a raw fetch (auth, tenant). */
  headers(): Record<string, string>;
  fetch: typeof fetch;
}

/** The wire body of a create (shared by the HTTP and WebSocket APIs). */
export function createRunBody(options: CreateAgentRunOptions): Record<string, unknown> {
  return {
    subject: options.subject,
    branch: options.branch,
    agent_ref: options.agentRef,
    create_key: options.createKey,
    budgets: options.budgets,
    input: options.input ?? null,
    reducer: options.reducer,
    executor_config: options.executorConfig,
    control_capability: options.controlCapability,
    as_agent: options.asAgent,
  };
}

/** What a run is about: `{workspace, path, node_id?}` or `"workspace:/path"`. */
export type AgentRunSubject = { workspace: string; path?: string; node_id?: string | null } | string;

/** `"ws:/path"` → `{workspace, path}`. */
export function subjectOf(subject: AgentRunSubject): { workspace: string; path?: string; node_id?: string | null } {
  if (typeof subject !== 'string') return subject;
  const cut = subject.indexOf(':/');
  if (cut <= 0) throw new Error(`subject must be "workspace:/path", got "${subject}"`);
  return { workspace: subject.slice(0, cut), path: subject.slice(cut + 1) };
}

let controlCounter = 0;

/** A fresh control id (controls are idempotent per id; retry with the SAME id). */
export function newControlId(prefix = 'ctl'): string {
  controlCounter += 1;
  const rand = Math.random().toString(36).slice(2, 10);
  return `${prefix}-${Date.now().toString(36)}-${controlCounter}-${rand}`;
}

export class AgentRunsApi {
  constructor(
    private repository: string,
    private transport: AgentRunsTransport,
  ) {}

  private base(runId?: string, rest = ''): string {
    const repo = encodeURIComponent(this.repository);
    const run = runId ? `/${encodeURIComponent(runId)}` : '';
    return `/api/agent-runs/${repo}${run}${rest}`;
  }

  /** Create a run, or get back the subject's live run. */
  create(options: CreateAgentRunOptions): Promise<CreateAgentRunResult> {
    return this.transport.request('POST', this.base(), createRunBody(options));
  }

  /** Every run about `subject` you may see, newest (the live one) first. */
  bySubject(subject: AgentRunSubject, limit = 20): Promise<AgentRunView[]> {
    const s = subjectOf(subject);
    const q = new URLSearchParams({ subject: `${s.workspace}:${s.path ?? ''}`, limit: String(limit) });
    if (s.node_id) q.set('subject_node_id', s.node_id);
    return this.transport.request('GET', `${this.base()}?${q.toString()}`);
  }

  /** Read a run (record + effective projection). */
  get(runId: string): Promise<AgentRunView> {
    return this.transport.request('GET', this.base(runId));
  }

  /** Your runs in one status (default "running"). */
  list(status: AgentRunStatus = 'running', limit = 50): Promise<AgentRunView[]> {
    return this.transport.request('GET', `${this.base()}?status=${status}&limit=${limit}`);
  }

  /** Durable events after `afterSeq` (gap-free; page with the last seq). */
  events(runId: string, afterSeq = 0, limit = 500): Promise<AgentRunEvent[]> {
    return this.transport.request('GET', this.base(runId, `/events?after_seq=${afterSeq}&limit=${limit}`));
  }

  /** Any control. Retrying with the same `controlId` is a no-op. */
  control(
    runId: string,
    command: AgentRunCommand,
    controlId: string = newControlId(command.command),
    capability?: string,
  ): Promise<AgentRunControlAck> {
    return this.transport.request('POST', this.base(runId, '/control'), {
      control_id: controlId,
      command,
      capability,
    });
  }

  stop(runId: string, reason?: string, controlId?: string): Promise<AgentRunControlAck> {
    return this.control(runId, { command: 'stop', reason }, controlId);
  }
  pause(runId: string, controlId?: string): Promise<AgentRunControlAck> {
    return this.control(runId, { command: 'pause' }, controlId);
  }
  resume(runId: string, controlId?: string): Promise<AgentRunControlAck> {
    return this.control(runId, { command: 'resume' }, controlId);
  }
  /** Queue input for the run's next safe boundary. */
  steer(runId: string, input: unknown, controlId?: string): Promise<AgentRunControlAck> {
    return this.control(runId, { command: 'steer', input }, controlId);
  }
  /** Decide an approval request; the digest ties it to one changeset. */
  approve(
    runId: string,
    requestId: string,
    subjectDigest: string,
    reject?: { reason?: string },
    controlId?: string,
  ): Promise<AgentRunControlAck> {
    const decision = reject ? { decision: 'reject' as const, reason: reject.reason } : { decision: 'approve' as const };
    return this.control(
      runId,
      { command: 'approve', request_id: requestId, decision, subject_digest: subjectDigest },
      controlId,
    );
  }
  /** Answer an input request. */
  answer(runId: string, requestId: string, value: unknown, controlId?: string): Promise<AgentRunControlAck> {
    return this.control(runId, { command: 'provide_input', request_id: requestId, value }, controlId);
  }

  /**
   * Stream the run's events: replay after `afterSeq`, then live. Ends after
   * the run's terminal event. Resume after a disconnect with the last `seq`.
   */
  async *stream(runId: string, afterSeq = 0, signal?: AbortSignal): AsyncGenerator<AgentRunEvent> {
    const response = await this.transport.fetch(
      this.transport.url(this.base(runId, `/stream?after_seq=${afterSeq}`)),
      { headers: { ...this.transport.headers(), Accept: 'text/event-stream' }, signal },
    );
    if (!response.ok || !response.body) {
      throw new Error(`agent run stream failed: ${response.status} ${await response.text()}`);
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    for (;;) {
      const { value, done } = await reader.read();
      if (done) return;
      buffer += decoder.decode(value, { stream: true });
      let cut: number;
      while ((cut = buffer.indexOf('\n\n')) >= 0) {
        const block = buffer.slice(0, cut);
        buffer = buffer.slice(cut + 2);
        let event = 'message';
        let data = '';
        for (const line of block.split('\n')) {
          if (line.startsWith('event:')) event = line.slice(6).trim();
          else if (line.startsWith('data:')) data += line.slice(5).trim();
        }
        if (event === 'end') return;
        if (event === 'error') throw new Error(`agent run stream error: ${data}`);
        if (event === 'run-event' && data) yield JSON.parse(data) as AgentRunEvent;
      }
    }
  }

  // ---- client-driven runs -------------------------------------------------

  /** Take the lease of a queued client-driven run. */
  acquire(runId: string): Promise<AgentRunFence> {
    return this.transport.request('POST', this.base(runId, '/lease/acquire'), {});
  }
  renew(runId: string, fence: AgentRunFence): Promise<AgentRunStatus> {
    return this.transport.request('POST', this.base(runId, '/lease/renew'), { fence });
  }
  release(runId: string, fence: AgentRunFence): Promise<AgentRunStatus> {
    return this.transport.request('POST', this.base(runId, '/lease/release'), { fence });
  }
  /** Begin an operation; returns the active operation (its `op_id`). */
  begin(runId: string, fence: AgentRunFence, op: BeginOperationOptions): Promise<{ op_id: string } & Record<string, unknown>> {
    return this.transport.request('POST', this.base(runId, '/operations'), { fence, ...op });
  }
  /** Record an operation's result. */
  finish(runId: string, fence: AgentRunFence, opId: string, result: FinishOperationOptions): Promise<AgentRunStatus> {
    return this.transport.request('POST', this.base(runId, '/operations/finish'), { fence, op_id: opId, ...result });
  }
  /** Open approval/input requests and wait (`kind: "approval" | "input"`). */
  wait(runId: string, fence: AgentRunFence, requests: Array<Record<string, unknown>>): Promise<AgentRunStatus> {
    return this.transport.request('POST', this.base(runId, '/wait'), { fence, requests });
  }
  /** End a client-driven run. */
  complete(
    runId: string,
    fence: AgentRunFence,
    status: 'completed' | 'failed',
    outcome: { kind: string; code?: string; message?: string; detail?: unknown },
  ): Promise<AgentRunStatus> {
    return this.transport.request('POST', this.base(runId, '/complete'), { fence, status, outcome });
  }

  // ── Child runs, mailbox, checkpoints, usage ──────────────────────────────

  /** Spawn a child with a typed objective; its budgets come out of this run's. */
  spawnChild(runId: string, options: SpawnChildOptions): Promise<SpawnChildResult> {
    const q = options.branch ? `?branch=${encodeURIComponent(options.branch)}` : '';
    return this.transport.request('POST', this.base(runId, `/children${q}`), spawnBody(options));
  }
  /** Every child of a run: the parent's link plus the child's live status. */
  children(runId: string): Promise<Array<{ link: Record<string, unknown>; status?: AgentRunStatus; usage?: unknown }>> {
    return this.transport.request('GET', this.base(runId, '/children'));
  }
  /** Inspect one child: record, events after `afterSeq`, usage, checkpoint. */
  inspectChild(runId: string, childId: string, afterSeq = 0, limit = 200): Promise<Record<string, unknown>> {
    const rest = `/children/${encodeURIComponent(childId)}?after_seq=${afterSeq}&limit=${limit}`;
    return this.transport.request('GET', this.base(runId, rest));
  }
  /** Message, steer, interrupt or resume a child (idempotent per `controlId`). */
  controlChild(
    runId: string,
    childId: string,
    action: ChildAction,
    controlId: string = newControlId(action.action),
  ): Promise<AgentRunControlAck> {
    const rest = `/children/${encodeURIComponent(childId)}/control`;
    return this.transport.request('POST', this.base(runId, rest), { control_id: controlId, ...action });
  }
  messageChild(runId: string, childId: string, message: unknown, controlId?: string): Promise<AgentRunControlAck> {
    return this.controlChild(runId, childId, { action: 'message', message }, controlId);
  }
  steerChild(runId: string, childId: string, input: unknown, controlId?: string): Promise<AgentRunControlAck> {
    return this.controlChild(runId, childId, { action: 'steer', input }, controlId);
  }
  interruptChild(runId: string, childId: string, reason?: string, mode: 'stop' | 'pause' = 'stop', controlId?: string): Promise<AgentRunControlAck> {
    return this.controlChild(runId, childId, { action: 'interrupt', mode, reason }, controlId);
  }
  /** A client-driven parent waits (under its lease) for a child's hand-back. */
  waitChild(runId: string, childId: string, fence: AgentRunFence, expiresAtMs?: number): Promise<AgentRunStatus> {
    const rest = `/children/${encodeURIComponent(childId)}/wait`;
    return this.transport.request('POST', this.base(runId, rest), { fence, expires_at_ms: expiresAtMs });
  }
  /** Unacknowledged mailbox items (hand-backs and child messages). */
  mailbox(runId: string): Promise<MailboxEntry[]> {
    return this.transport.request('GET', this.base(runId, '/mailbox'));
  }
  /** Acknowledge mailbox items up to `upTo`. */
  ackMailbox(runId: string, upTo: number): Promise<{ remaining: number }> {
    return this.transport.request('POST', this.base(runId, '/mailbox/ack'), { up_to: upTo });
  }
  /** A child posts a message to its parent's mailbox (idempotent per id). */
  postToParent(runId: string, messageId: string, message: unknown): Promise<{ mail_no: number }> {
    return this.transport.request('POST', this.base(runId, '/post-to-parent'), { message_id: messageId, message });
  }
  /** Write a structured checkpoint (compaction keeps state, not prose). */
  checkpoint(runId: string, options: CheckpointOptions): Promise<Record<string, unknown>> {
    return this.transport.request('POST', this.base(runId, '/checkpoints'), checkpointBody(options));
  }
  /** Read a checkpoint (default: latest) with its structured state. */
  readCheckpoint(runId: string, checkpointNo?: number): Promise<{ checkpoint: Record<string, unknown>; state: unknown }> {
    return this.transport.request('GET', this.base(runId, `/checkpoints/${checkpointNo ?? 'latest'}`));
  }
  /** Usage accounting: own, children, reserved, spare. */
  usage(runId: string): Promise<Record<string, unknown>> {
    return this.transport.request('GET', this.base(runId, '/usage'));
  }
}

File: crates/raisin-transport-http/src/handlers/agent_runs.rs
// SPDX-License-Identifier: BSL-1.1

//! Durable agent runs over HTTP.
//!
//! Thin adapters over `raisin_agent_runtime::api`: this layer authenticates,
//! maps the caller, and serializes; every rule (who may see or control a run,
//! lease fencing, idempotency) is core's. The same shapes are spoken by the
//! function bindings and the SDKs, so an external agent can drive a run with
//! nothing but this API.
//!
//! ```text
//! POST /api/agent-runs/{repo}                         create (or find) a run
//! GET  /api/agent-runs/{repo}?status=&branch=&limit=  list visible runs
//! GET  /api/agent-runs/{repo}?subject=ws:/path        every run of a subject,
//!                                                     newest (live) first
//! GET  /api/agent-runs/{repo}/{run}?branch=           read (with projection)
//! GET  /api/agent-runs/{repo}/{run}/events?after_seq= durable events
//! GET  /api/agent-runs/{repo}/{run}/stream?after_seq= SSE: replay, then live
//! POST /api/agent-runs/{repo}/{run}/control           any control command
//! POST /api/agent-runs/{repo}/{run}/{stop|pause|resume|steer|approve|answer}
//! POST /api/agent-runs/{repo}/{run}/lease/{acquire|renew|release}
//! POST /api/agent-runs/{repo}/{run}/operations        begin (client driver)
//! POST /api/agent-runs/{repo}/{run}/operations/finish finish (body: op_id)
//! POST /api/agent-runs/{repo}/{run}/{wait|complete}   client driver
//! ```

use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::{Extension, Json};
use raisin_agent_runtime::api::{self, Caller};
use raisin_agent_runtime::control::{ApprovalDecision, ControlKind};
use raisin_agent_runtime::host::AgentRunHost;
use raisin_agent_runtime::ids::{RequestId, RunId, RunScope, SubjectRef};
use raisin_agent_runtime::state::RunStatus;
use raisin_models::auth::AuthContext;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::error::ApiError;
use crate::middleware::TenantInfo;
use crate::state::AppState;

/// `?branch=`.
#[derive(Debug, Deserialize, Default)]
pub struct BranchQuery {
    /// Branch (default `main`).
    pub branch: Option<String>,
    /// Events after this seq.
    pub after_seq: Option<u64>,
    /// Page size.
    pub limit: Option<usize>,
    /// Status filter (list).
    pub status: Option<String>,
    /// Subject filter (list): `"{workspace}:{path}"`.
    pub subject: Option<String>,
    /// The subject's node id, when the caller knows it.
    pub subject_node_id: Option<String>,
}

pub(crate) fn host() -> Result<std::sync::Arc<AgentRunHost>, ApiError> {
    raisin_rocksdb::agent_runs::agent_run_host().ok_or_else(|| {
        ApiError::new(
            StatusCode::SERVICE_UNAVAILABLE,
            "agent_runs_unavailable",
            "the agent run runtime is not running on this server",
        )
    })
}

pub(crate) fn map_err(e: api::ApiError) -> ApiError {
    let status = StatusCode::from_u16(e.status).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
    ApiError::new(status, e.code, e.message)
}

/// The authenticated caller; anonymous callers are refused.
pub(crate) fn caller(auth: Option<Extension<AuthContext>>) -> Result<Caller, ApiError> {
    let Some(Extension(auth)) = auth else {
        return Err(ApiError::unauthorized("authentication required"));
    };
    if auth.is_system {
        return Ok(Caller {
            id: "system".into(),
            admin: true,
        });
    }
    if auth.is_anonymous_principal() {
        return Err(ApiError::unauthorized("authentication required"));
    }
    let id = auth.user_id.clone().unwrap_or_default();
    Ok(Caller { id, admin: false })
}

pub(crate) fn scope_of(tenant: &TenantInfo, repo: &str, q: &BranchQuery) -> RunScope {
    api::scope(&tenant.tenant_id, repo, q.branch.as_deref())
}

type Resp = Result<Json<Value>, ApiError>;

fn ok<T: serde::Serialize>(v: T) -> Resp {
    Ok(Json(serde_json::to_value(v).unwrap_or(Value::Null)))
}

/// Create (or find) a run.
pub async fn create_run(
    Extension(tenant): Extension<TenantInfo>,
    Path(repo): Path<String>,
    auth: Option<Extension<AuthContext>>,
    Json(req): Json<api::CreateRunRequest>,
) -> Resp {
    let caller = caller(auth)?;
    ok(
        api::create(&*host()?, &tenant.tenant_id, &repo, &caller, req)
            .await
            .map_err(map_err)?,
    )
}

fn parse_status(s: Option<&str>) -> Result<RunStatus, ApiError> {
    let s = s.unwrap_or("running");
    serde_json::from_value(json!(s))
        .map_err(|_| ApiError::validation_failed(format!("unknown status '{s}'")))
}

/// List the caller's runs: of one subject (newest first), or in one status.
pub async fn list_runs(
    State(state): State<AppState>,
    Extension(tenant): Extension<TenantInfo>,
    Path(repo): Path<String>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
) -> Resp {
    let caller = caller(auth)?;
    let limit = q.limit.unwrap_or(50).min(500);
    let scope = scope_of(&tenant, &repo, &q);
    if let Some(subject) = q.subject.as_deref() {
        let subjects =
            subject_spellings(&state, &scope, subject, q.subject_node_id.clone()).await?;
        let mut runs = api::by_subject(&*host()?, &scope, &caller, &subjects, limit)
            .await
            .map_err(map_err)?;
        if let Some(s) = q.status.as_deref() {
            let status = parse_status(Some(s))?;
            runs.retain(|v| v.status == status);
        }
        return ok(runs);
    }
    let status = parse_status(q.status.as_deref())?;
    ok(api::list(&*host()?, &scope, &caller, status, limit)
        .await
        .map_err(map_err)?)
}

/// `"{workspace}:{path}"` in both spellings a run may be keyed by: its node id
/// (when the creator knew it) and its path.
async fn subject_spellings(
    state: &AppState,
    scope: &RunScope,
    subject: &str,
    node_id: Option<String>,
) -> Result<Vec<SubjectRef>, ApiError> {
    use raisin_storage::{NodeRepository, Storage};
    let (workspace, path) = subject
        .split_once(':')
        .filter(|(w, p)| !w.is_empty() && p.starts_with('/'))
        .ok_or_else(|| ApiError::validation_failed("subject must be '{workspace}:{path}'"))?;
    let node_id = match node_id {
        Some(id) => Some(id),
        None => {
            let s = raisin_storage::scope::StorageScope::new(
                &scope.tenant_id,
                &scope.repo_id,
                &scope.branch,
                workspace,
            );
            state
                .storage
                .nodes()
                .get_by_path(s, path, None)
                .await
                .ok()
                .flatten()
                .map(|n| n.id)
        }
    };
    let mut out = Vec::new();
    if let Some(id) = node_id {
        out.push(SubjectRef {
            workspace: workspace.into(),
            path: path.into(),
            node_id: Some(id),
        });
    }
    out.push(SubjectRef {
        workspace: workspace.into(),
        path: path.into(),
        node_id: None,
    });
    Ok(out)
}

/// Read a run.
pub async fn get_run(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run)): Path<(String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    ok(api::get(&*host()?, &scope, &RunId(run), &caller)
        .await
        .map_err(map_err)?)
}

/// Durable events after `after_seq`.
pub async fn run_events(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run)): Path<(String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    let limit = q.limit.unwrap_or(500).min(5_000);
    let events = api::events(
        &*host()?,
        &scope,
        &RunId(run),
        &caller,
        q.after_seq.unwrap_or(0),
        limit,
    )
    .await
    .map_err(map_err)?;
    ok(events)
}

/// Any control command.
pub async fn control_run(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run)): Path<(String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
    Json(req): Json<api::ControlRequest>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    ok(api::control(&*host()?, &scope, &RunId(run), &caller, req)
        .await
        .map_err(map_err)?)
}

/// Body of the named-action shortcuts.
#[derive(Debug, Deserialize)]
pub struct ActionBody {
    /// Idempotency key.
    pub control_id: String,
    /// Capability secret, for a non-principal controller.
    #[serde(default)]
    pub capability: Option<String>,
    /// stop: reason; approve: rejection reason.
    #[serde(default)]
    pub reason: Option<String>,
    /// steer: the input.
    #[serde(default)]
    pub input: Option<Value>,
    /// approve / answer: the request.
    #[serde(default)]
    pub request_id: Option<String>,
    /// approve: `approve` (default) or `reject`.
    #[serde(default)]
    pub decision: Option<String>,
    /// approve: the digest the approval is tied to.
    #[serde(default)]
    pub subject_digest: Option<String>,
    /// answer: the value.
    #[serde(default)]
    pub value: Option<Value>,
    /// resume: accept a changed reducer.
    #[serde(default)]
    pub accept_reducer_change: bool,
}

fn need<T>(v: Option<T>, field: &str) -> Result<T, ApiError> {
    v.ok_or_else(|| ApiError::missing_required_field(field))
}

fn action_command(action: &str, b: &ActionBody) -> Result<ControlKind, ApiError> {
    Ok(match action {
        "stop" => ControlKind::Stop {
            reason: b.reason.clone(),
        },
        "pause" => ControlKind::Pause,
        "resume" => ControlKind::Resume {
            budget_increase: None,
            accept_reducer_change: b.accept_reducer_change,
        },
        "steer" => ControlKind::Steer {
            input: need(b.input.clone(), "input")?,
        },
        "approve" => ControlKind::Approve {
            request_id: RequestId(need(b.request_id.clone(), "request_id")?),
            decision: match b.decision.as_deref() {
                Some("reject") => ApprovalDecision::Reject {
                    reason: b.reason.clone(),
                },
                _ => ApprovalDecision::Approve,
            },
            subject_digest: need(b.subject_digest.clone(), "subject_digest")?,
        },
        "answer" => ControlKind::ProvideInput {
            request_id: RequestId(need(b.request_id.clone(), "request_id")?),
            value: need(b.value.clone(), "value")?,
        },
        other => return Err(ApiError::not_found(format!("unknown run action '{other}'"))),
    })
}

/// `stop | pause | resume | steer | approve | answer | wait | complete`.
pub async fn run_action(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run, action)): Path<(String, String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
    Json(body): Json<Value>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    let host = host()?;
    let run = RunId(run);
    let bad = |e: serde_json::Error| ApiError::validation_failed(e.to_string());
    match action.as_str() {
        "wait" => ok(api::wait(
            &host,
            &scope,
            &run,
            &caller,
            serde_json::from_value(body).map_err(bad)?,
        )
        .await
        .map_err(map_err)?),
        "complete" => ok(api::complete(
            &host,
            &scope,
            &run,
            &caller,
            serde_json::from_value(body).map_err(bad)?,
        )
        .await
        .map_err(map_err)?),
        _ => {
            let b: ActionBody = serde_json::from_value(body).map_err(bad)?;
            let req = api::ControlRequest {
                control_id: b.control_id.clone(),
                command: action_command(&action, &b)?,
                capability: b.capability.clone(),
            };
            ok(api::control(&host, &scope, &run, &caller, req)
                .await
                .map_err(map_err)?)
        }
    }
}

/// `acquire | renew | release` the lease (client-driven runs).
pub async fn run_lease(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run, action)): Path<(String, String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
    body: Option<Json<Value>>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    let host = host()?;
    let run = RunId(run);
    let body = body.map(|Json(v)| v).unwrap_or(Value::Null);
    let fence = || -> Result<api::Fence, ApiError> {
        serde_json::from_value(body.get("fence").cloned().unwrap_or(Value::Null))
            .map_err(|_| ApiError::missing_required_field("fence"))
    };
    match action.as_str() {
        "acquire" => {
            let owner = format!("client:{}:{}", caller.id, uuid::Uuid::new_v4());
            ok(api::acquire(&host, &scope, &run, &caller, &owner)
                .await
                .map_err(map_err)?)
        }
        "renew" => ok(api::renew(&host, &scope, &run, &caller, &fence()?)
            .await
            .map_err(map_err)?),
        "release" => ok(api::release(&host, &scope, &run, &caller, &fence()?)
            .await
            .map_err(map_err)?),
        other => Err(ApiError::not_found(format!(
            "unknown lease action '{other}'"
        ))),
    }
}

/// Begin an operation (client-driven runs).
pub async fn begin_operation(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run)): Path<(String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
    Json(req): Json<api::BeginRequest>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    ok(api::begin(&*host()?, &scope, &RunId(run), &caller, req)
        .await
        .map_err(map_err)?)
}

/// `{ op_id, fence, outcome, payload?, tool_calls?, usage?, resume_key? }`.
#[derive(Debug, Deserialize)]
pub struct FinishBody {
    /// The operation.
    pub op_id: String,
    /// The result.
    #[serde(flatten)]
    pub result: api::FinishRequest,
}

/// Finish an operation (client-driven runs).
pub async fn finish_operation(
    Extension(tenant): Extension<TenantInfo>,
    Path((repo, run)): Path<(String, String)>,
    Query(q): Query<BranchQuery>,
    auth: Option<Extension<AuthContext>>,
    Json(body): Json<FinishBody>,
) -> Resp {
    let caller = caller(auth)?;
    let scope = scope_of(&tenant, &repo, &q);
    ok(api::finish(
        &*host()?,
        &scope,
        &RunId(run),
        &caller,
        &body.op_id,
        body.result,
    )
    .await
    .map_err(map_err)?)
}
